### PR TITLE
[nozomi-networks-1484] Nozomi networks integration

### PR DESCRIPTION
**Pull request by nicoloereni**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
First release
Integration implementation with fetch incidents and some commands

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation 

## Demisto Partner?
- [X] **nozomi-networks-1484 integration**

### CODE REVIEW

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      type: String
+    - contextPath: type
+      description: asset type as 'OT_device'
+      type: String
+    - contextPath: protocols
+      description: array of asset protocols
+      type: Unknown
+  - arguments:
+    - default: false
+      defaultValue: '[]'
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:150]: remove

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 2]
```diff
@@ -0,0 +1,253 @@
+category: Network Security
+commonfields:
+  id: Nozomi Networks
+  version: -1
+configuration:
+- additionalinfo: Here you have to insert the endpoint url of your Nozomi Networks
+    appliance
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:7]:
```suggestion
- additionalinfo: Your Nozomi Networks appliance endpoint URL
   
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  name: isFetch
+  required: false
+  type: 8
+- display: Incident type
+  name: incidentType
+  required: false
+  type: 13
+- additionalinfo: 'Here you can decide from how many time from the past to get incidents.
+    Default is 7 days. some examples of valid input are: 30 seconds, 1 minute, 4 hours,
+    2 days...'
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:43]:
```suggestion
- additionalinfo: 'Set the time frame to get incidents by.
    Default is 7 days. Other examples: 30 seconds, 1 minute, 4 hours,
    2 days...'
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  defaultvalue: 7 days
+  display: Get incidents from last
+  hidden: false
+  name: fetchTime
+  required: false
+  type: 0
+- additionalinfo: Get alert having risk higher or equal the one selected.
+  defaultvalue: '1'
+  display: Get alerts from risk
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:52]:
```suggestion
  display: Minimal risk to fetch incidents.
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+- additionalinfo: 'Here you can decide from how many time from the past to get incidents.
+    Default is 7 days. some examples of valid input are: 30 seconds, 1 minute, 4 hours,
+    2 days...'
+  defaultvalue: 7 days
+  display: Get incidents from last
+  hidden: false
+  name: fetchTime
+  required: false
+  type: 0
+- additionalinfo: Get alert having risk higher or equal the one selected.
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:50]:
```suggestion
- additionalinfo: Will only fetch incident with a risk higher or equal to the selected value.
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+  - '8'
+  - '9'
+  - '10'
+  required: false
+  type: 15
+- additionalinfo: In Nozomi platform there are alerts as a single events and there
+    is also the concept of incident, that is an aggegation of alert, by default we
+    just send alerts to demisto, if you flag this checkbox we also send incidents.
+    When a fetch integration was configured without incidents, if enabled means that
+    the new n2os incidents will be sent to XSOAR, but not the old one.
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:72]:
```suggestion
- additionalinfo: In Nozomi platform, there are alerts as a single events and there
    are incidents, which are an aggregation of alerts. By default only alerts will be fetched to CORTEX XSOAR, however, if selected, also incidents will be fetched.
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+  display: Fetch also Nozomi incidents
+  hidden: false
+  name: fecthAlsoIncidents
+  required: false
+  type: 8
+description: |-
+  The Nozomi Networks Guardian platform is a hardware or virtual appliance that is used to monitor OT/IoT/IT networks. It combines asset discovery, network visualization, vulnerability assessment, risk monitoring and threat detection in a single solution.
+    This integration is used to gather alert and asset information from Nozomi.
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:80]:
```suggestion
    This integration is used to gather alerts and assets information from Nozomi.
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+display: Nozomi Networks
+name: Nozomi Networks
+script:
+  commands:
+  - arguments:
+    - default: false
+      description: You can add a filter to get exactly the assets you want. For example
+        'where ip match 10.0.1.10', 'where vendor ==  Selta Telematica S.p.a'
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:88]: Let's mention here that the filer needs to be taken from the Nozomi UI.

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+
+    ibtt = incidents_better_than_time(st, head, risk, also_n2os_incidents, client)
+
+    lft = last_fetched_time(ibtt, last_run)
+    lfid = last_fetched_id(ibtt, last_run)
+
+    incidents_merged = incidents_equal_time_better_id(st, last_id, risk, also_n2os_incidents, client) + ibtt
+
+    parsed_incidents = [parse_incident(i) for i in incidents_merged]
+    parsed_incidents.sort(key=get_incident_name)
```
[the remark this answers is not in the document]
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:176]: cause they can have the same time and it is common, so it's not enough to have the last time but we need also the last id, so we use id as second sorting element to get the alerts with the same time but better id not yet sent to demisto.

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+        return_results(CommandResults(**close_incidents_as_security(demisto.args())))
+    elif demisto.command() == 'nozomi-find-assets':
+        return_results(CommandResults(**find_assets(demisto.args())))
+    elif demisto.command() == 'nozomi-query':
+        return_results(CommandResults(**query(demisto.args())))
+    elif demisto.command() == 'nozomi-find-ip-by-mac':
+        return_results(CommandResults(**find_ip_by_mac(demisto.args())))
+except Exception as e:
+    LOG(f'nozomi: got an error {e}')
+    LOG.print_log()
```
[the remark this answers is not in the document]
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:482]: putting it in 
def main():

seems not to work

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml (hunk 1 above)
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:150]: all the four lines?
Sorry I dont understand :)

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:482]: btw there is a try catch so for me it's the same

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml (hunk 8 above)
Comment (ShahafBenYakir) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:77]: ?

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+):
+    st = st or start_time(demisto.getLastRun(), demisto.params().get('fetchTime', '7 days').strip())
+    client = client or get_client()
+    last_run = last_run or demisto.getLastRun()
+    last_id = last_id or get_last_id(demisto.getLastRun())
+    risk = risk or demisto.params().get('riskFrom', None)
+    fetch_also_n2os_incidents = fetch_also_n2os_incidents or demisto.params().get('fecthAlsoIncidents', False)
+
+    if test_mode:
+        return [], None  # cause we want just field validation from test fetch
```
[the remark this answers is not in the document]
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py]: I fix this, 
in case of test mode I don't execute even ack_incidents to not chenge nozomi platform alerts state

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+):
+    st = st or start_time(demisto.getLastRun(), demisto.params().get('fetchTime', '7 days').strip())
+    last_run = last_run or demisto.getLastRun()
+    last_id = last_id or get_last_id(demisto.getLastRun())
+    risk = risk or demisto.params().get('riskFrom', None)
+    fetch_also_n2os_incidents = fetch_also_n2os_incidents or demisto.params().get('fecthAlsoIncidents', False)
+
+    demisto_incidents, last_fetch, last_id_returned = \
+        incidents(st, last_id, last_run, risk, fetch_also_n2os_incidents, client)
+    demisto.setLastRun({'last_fetch': last_fetch, 'last_id': last_id_returned})
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:296]: Isn't this also should be part of the `if not test_mode:`? We don't want to change the last run if we are only testing.

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+
+
+def fetch_incidents(
+        client,
+        st=None,
+        last_run=None,
+        last_id=None,
+        risk=None,
+        fetch_also_n2os_incidents=None,
+        test_mode=False
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:286]: Why is this needed?

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+    if not test_mode:
+        ack_alerts(nozomi_alerts_ids_from_demisto_incidents(demisto_incidents), client)
+        demisto.incidents(demisto_incidents)
+
+    return demisto_incidents, last_fetch
+
+
+def is_alive(client):
+    client = client or get_client()
+    error = None
```
Comment (ShahafBenYakir) [inline Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:307]: Any reason why `client` could be None?

```suggestion
def is_alive(client):
    error = None
```

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py (hunk 13 above)
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:286]: To test the function with different values

Comment (ShahafBenYakir) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:286]: But I don't see we use this anywhere. Plus this values will always come from the user no? Which is exactly what is happening in the following lines.

Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:286]: I have integration test, not sent to the PR cause they need a running instance of nozomi to run.
Something like that:

`def test_fetch_incidents_without_n2os_incidents():
    i, last_fetch = fetch_incidents(
        __real_client(),
        st='1392040975219',
        risk='1',
        fetch_also_n2os_incidents=False,
    )
    assert last_fetch is not None
    assert len(i) == 16`

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml (hunk 8 above)
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.yml:77]: You can use the filter in the nozomi UI as well as in demisto, nozomi ui does not have already done filters, so it's strange IMHO to ask the user to take filter from there.
The user have to know how to filter elements in nozomi platform, so I add

File: Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nicoloereni) [reply Packs/NozomiNetworks/Integrations/NozomiNetworks/NozomiNetworks.py:175]: yes exactly
